hash/hash_blake2b.c

```diff
@@ -0,0 +1,378 @@
+
+                bytes[j] = (h[i / 8] & (mask)) >> offset;
+            }
+        }
+
+        dest[i] = bytes[i % 8];
+    }
+
+    return 0;
+}
+
+/* @brief blake2b hash function
+ *
+ * @param message the message to be hashed
+ * @param len length of message (0 <= len < 2**128) (depends on sizeof(size_t)
+ * for this implementation)
+ * @param key optional secret key
+ * @param kk length of optional secret key (0 <= kk <= 64)
+ * @param nn length of output digest (1 <= nn < 64)
+ *
+ * @returns NULL if heap memory couldn't be allocated. Otherwise heap allocated
+ * memory nn bytes large
+ */
+uint8_t *blake2b(const uint8_t *message, size_t len, const uint8_t *key,
+                 uint8_t kk, uint8_t nn)
+{
+    uint8_t *dest = NULL;
+    uint64_t(*blocks)[16], ll[2], long_hold;
+    size_t dd, has_key, i;
+    size_t block_index, word_in_block;
+
+    if (message == NULL)
+        len = 0;
+    if (key == NULL)
+        kk = 0;
+
+    kk = MIN(kk, 64);
+    nn = MIN(nn, 64);
+
+    dd = MAX(CEIL(kk, bb) + CEIL(len, bb), 1);
+
+    blocks = calloc(dd, sizeof(uint64_t[16]));
+    if (blocks == NULL)
+        return NULL;
+
+    dest = malloc(nn * sizeof(uint8_t));
+    if (dest == NULL)
+    {
+        free(blocks);
+        return NULL;
+    }
+
+    /* If there is a secret key it occupies the first block */
+    for (i = 0; i < kk; i++)
+    {
+        long_hold = message[i];
+        long_hold <<= 8 * (i % 8);
+
+        word_in_block = (i % bb) / 8;
+        /* block_index will always be 0 because kk <= 64 and bb = 128*/
+        blocks[0][word_in_block] |= long_hold;
+    }
+
+    has_key = kk > 0 ? 1 : 0;
+
+    for (i = 0; i < len; i++)
+    {
+        /* long_hold exists because the bit-shifting will overflow if we don't
+         * store the value */
+        long_hold = message[i];
+        long_hold <<= 8 * (i % 8);
+
+        block_index = has_key + (i / bb);
+        word_in_block = (i % bb) / 8;
+
+        blocks[block_index][word_in_block] |= long_hold;
+    }
+
+    ll[0] = len & UINT64_MAX;
+
+    /* The C standard does not specify a maximum length for size_t,
+     * although most machines implement it to be the same length as uint64_t.
+     * On machines where size_t is 8 bytes long this will issue a compiler 
+     * warning, which is why it is suppressed. But on a machine where size_t
+     * is greater than 8 bytes, this will work as normal. */
+    if (sizeof(len) > 8)
+    {
+        ll[1] = len >> 64;
+    }
+    else
+    {
+        ll[1] = 0;
+    }
+
+    BLAKE2B(dest, blocks, dd, ll, kk, nn);
+
+    return dest;
+}
+
+/** @} */
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void assert_bytes(const uint8_t *expected, const uint8_t *actual,
+                         uint8_t len)
+{
+    uint8_t i;
+
+    assert(expected != NULL);
+    assert(actual != NULL);
+    assert(len > 0);
+
+    for (i = 0; i < len; i++)
+    {
+        assert(expected[i] == actual[i]);
+    }
```

Comment:
A small printing message might help the users when testing/editing the algorithm.
```suggestion
    }

    printf("All tests have successfully passed!\n");
```